controllers/progressiverollout_controller.go

```diff
@@ -61,5 +69,93 @@ func (r *ProgressiveRolloutReconciler) Reconcile(ctx context.Context, req ctrl.R
 func (r *ProgressiveRolloutReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&deploymentskyscannernetv1alpha1.ProgressiveRollout{}).
+		Watches(
+			&source.Kind{Type: &argov1alpha1.Application{}},
+			handler.EnqueueRequestsFromMapFunc(r.requestsForApplicationChange)).
+		Watches(
+			&source.Kind{Type: &corev1.Secret{}}, handler.EnqueueRequestsFromMapFunc(r.requestsForSecretChange)).
 		Complete(r)
 }
+
+func (r *ProgressiveRolloutReconciler) requestsForApplicationChange(o client.Object) []reconcile.Request {
+
+	/*
+		We trigger a reconciliation loop on an Application event if:
+		- the Application owner is referenced by a ProgressiveRollout object
+	*/
+
+	var requests []reconcile.Request
+	var list deploymentskyscannernetv1alpha1.ProgressiveRolloutList
+	ctx := context.Background()
+
+	app, ok := o.(*argov1alpha1.Application)
+	if !ok {
+		err := fmt.Errorf("expected application, got %T", o)
+		r.Log.Error(err, "failed to convert object to application")
+		return nil
+	}
+
+	if err := r.List(ctx, &list); err != nil {
+		r.Log.Error(err, "failed to list ProgressiveRollout")
+		return nil
+	}
+
+	for _, pr := range list.Items {
+		if pr.HasOwnerReference(app.GetOwnerReferences()) {
+			requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+				Namespace: pr.Namespace,
+				Name:      pr.Name,
+			}})
+		}
+	}
+
+	return requests
+}
+
+func (r *ProgressiveRolloutReconciler) requestsForSecretChange(o client.Object) []reconcile.Request {
+
+	/*
+		We trigger a reconciliation loop on a Secret event if:
+		- the Secret is an ArgoCD cluster, AND
+		- there is an Application targeting that secret/cluster, AND
+		- that Application owner is referenced by a ProgressiveRollout object
+	*/
+
+	var requests []reconcile.Request
+	var prList deploymentskyscannernetv1alpha1.ProgressiveRolloutList
+	var appList argov1alpha1.ApplicationList
+	ctx := context.Background()
+
+	s, ok := o.(*corev1.Secret)
+	if !ok {
+		err := fmt.Errorf("expected secret, got %T", o)
+		r.Log.Error(err, "failed to convert object to secret")
+		return nil
+	}
+
+	if !internal.IsArgoCDCluster(s.GetLabels()) {
+		return nil
+	}
+
+	if err := r.List(ctx, &prList); err != nil {
+		r.Log.Error(err, "failed to list ProgressiveRollout")
+		return nil
+	}
+	if err := r.List(ctx, &appList); err != nil {
```

Comment:
❓ This function must be magic? It looks at the type of the objected passed it and lists thoses? 🤯

Reply:
The object type is `somethingList`. It works the same as https://github.com/Skyscanner/argocd-progressive-rollout/blob/main/controllers/progressiverollout_controller.go#L46